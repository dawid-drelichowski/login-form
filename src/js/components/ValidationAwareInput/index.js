// @flow
import * as React from 'react'

export default class ValidationAwareInput extends React.Component<
  {type: string, id: string, label: string, onChange?: (id: string, value: string) => void},
  {value: string, checked: boolean, invalidMessage: string}
> {
  state = {value: '', checked: false, invalidMessage: ''}
  onChange = (event: SyntheticInputEvent<HTMLInputElement>): void => {
    const onChange: (id: string, value: string) => void = this.props.onChange ? this.props.onChange : () => {},
      target: HTMLInputElement = event.target,
      value: string = target.value

    this.setState({value, checked: target.checked, invalidMessage: ''})
    onChange(target.id, value)
  }
  onInvalid = (event: SyntheticInputEvent<HTMLInputElement>): void => {
    this.setState({invalidMessage: `Invalid ${this.props.label.toLowerCase()}`})
    event.preventDefault()
  }
  render (): React.Node {
    const {type, id, label, onChange, ...props} = this.props,
      invalidMessage: string = this.state.invalidMessage

    return <div>
      <label htmlFor={id}>{label}</label>
      <input type={type} id={id} value={this.state.value} checked={this.state.checked} {...props}
        onChange={this.onChange} onInvalid={this.onInvalid}/>
      {invalidMessage ? <div>{invalidMessage}</div> : ''}
    </div>
  }
}
