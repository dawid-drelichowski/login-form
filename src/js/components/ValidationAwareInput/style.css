.validation-aware-input .input {
  background: #f2f2f2;
  width: 100%;
  border: 0;
  padding: 10px;
  display: block;
  box-shadow: none;
  box-sizing: border-box;
  font-size: 1rem;
  margin: 5px 0;
}

.validation-aware-input .invalid {
  color: #c54639;
  font-weight: bold;
  margin-bottom: 5px;
}
