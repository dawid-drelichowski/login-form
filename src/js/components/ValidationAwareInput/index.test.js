import React from 'react'
import ValidationAwareInput from './index'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'

describe('Validation aware input component', () => {
  function getRenderedComponent (type, id, label, props = {}) {
    return renderer.create(<ValidationAwareInput type={type} id={id} label={label} {...props}/>)
  }

  function getShallowComponent (type, id, label, onChange = () => {}, props = {}) {
    return shallow(<ValidationAwareInput type={type} id={id} label={label} onChange={onChange} {...props}/>)
  }

  it('should render without additional input attributes', () => {
    expect(getRenderedComponent('text', 'name', 'Your name').toJSON()).toMatchSnapshot()
  })

  it('should render with additional input attributes', () => {
    expect(getRenderedComponent('text', 'name', 'Your name', {placeholder: 'Your name', required: true})
      .toJSON()).toMatchSnapshot()
  })

  it('should have value when input changed', () => {
    const value = 'Dawid',
      inputSelector = 'input[type="text"]',
      form = getShallowComponent('text', 'name', 'Your name')

    form.find(inputSelector).simulate('change', {target: {value}})
    expect(form.find(inputSelector).prop('value')).toEqual(value)
  })
})
