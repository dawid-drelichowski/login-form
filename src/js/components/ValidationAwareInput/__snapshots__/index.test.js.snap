// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation aware input component should render with additional input attributes 1`] = `
<div>
  <label
    htmlFor="name"
  >
    Your name
  </label>
  <input
    checked={false}
    id="name"
    onChange={[Function]}
    onInvalid={[Function]}
    placeholder="Your name"
    required={true}
    type="text"
    value=""
  />
  
</div>
`;

exports[`Validation aware input component should render without additional input attributes 1`] = `
<div>
  <label
    htmlFor="name"
  >
    Your name
  </label>
  <input
    checked={false}
    id="name"
    onChange={[Function]}
    onInvalid={[Function]}
    type="text"
    value=""
  />
  
</div>
`;
